Blob migration steps documentation (#2145)
microsoft/dicom-server: docs/how-to-guides/blob-migration.md

# Blob migration

Currently DICOM files are stored with DICOM UIDs as blob names in blob storage, using the template `{account}/{container}/{studyUid}/{seriesUid}/{sopInstanceUid}_{watermark}.dcm`. 

> Since UIDs may include personal information about the context of their creation, such as patient information or identifiers. So we made the decision to change the way that we store DICOM files. In the next sections we have detailed steps for each scenario to migrate your existing blobs from old format to new format.

Reviewer: ```suggestion
Since UIDs may include personal information about the context of their creation, such as patient information or identifiers, we made the decision to change the way that we store DICOM files. In the next sections we list the steps to migrate your existing blobs from the old format to the new format.
```

## Blob migration configuration

> Below is the appsettings configuration for blob migration. Few properties needs to be updated to trigger migration.

Reviewer: ```suggestion
Below is the `appsettings.json` configuration related to blob migration. Several properties need to be updated to trigger migration.
```

```json
"DicomServer": {
    "Services": {
      "BlobMigration": {
        "FormatType": "Old",
        "StartCopy": false,
        "StartDelete": false,
        "CopyFileOperationId": "1d4689da-ca3b-4659-b0c7-7bf6c9ff25e1",
        "DeleteFileOperationId": "ce38a27e-b194-4645-b47a-fe91c38c330f",
        "CleanupDeletedFileOperationId": "d32a0469-9c27-4df3-a1e8-12f7f8fecbc8",
        "CleanupFilterTimeStamp": "2022-08-01"

>       }

Reviewer: ```suggestion
      }
    }
```

}
```

## Migration steps

> 1. If you are a new service and have not created any files, you can upgrade to the latest version of the service and skip the migration steps. That means you will see the BlobMigration.FormatType as "New" in appsettings configuration file. 

Reviewer: Once we bump the major version, we will remove the section of configuration, right?

```suggestion
1. If you are a new service and have not created any files, you can upgrade to the latest version of the service and skip the migration steps. The configuration section above will not be present.
```

> 2. If you have already uploaded DICOM files but does not care about previous data. You can use the [Delete API](../resources/conformance-statement.md#delete) to delete all the data which can free up space and then upgrade to the latest version of the service or you can upgrade to the latest version of the service and then use the [Delete API](../resources/conformance-statement.md#delete) to delete all the data.

Reviewer: It might be simpler to have the new major version only aware of the new format, so any old format needs to be removed before the upgrade.

```suggestion
2. If you have already uploaded DICOM files but do not care about migrating the data, you can use the [Delete API](../resources/conformance-statement.md#delete) to delete all existing studies. You can then upgrade to the latest version of the service.
```

> 2. If you have already uploaded DICOM files but does not care about previous data. You can use the [Delete API](../resources/conformance-statement.md#delete) to delete all the data which can free up space and then upgrade to the latest version of the service or you can upgrade to the latest version of the service and then use the [Delete API](../resources/conformance-statement.md#delete) to delete all the data.

Author: It's kind of optional, if they don't care about the data then it is going to fillup the storage and eventually will cost them.

> 3. If you have already uploaded DICOM files and want to keep the previous data, you need to execute the below steps to copy the files first. This scenario has two options depending on whether if you want interruption to the service or not. Make sure Azure monitor is configured to monitor the service before starting the migration, for more info on how to configure Azure monitor, please refer to [Azure Monitor](../how-to-guides/configure-dicom-server-settings.md#azure-monitor).

Reviewer: ```suggestion
3. If you have already uploaded DICOM files and want to migrate the data, you will need to execute the following steps before upgrading. This scenario has two options depending on whether if you want interruption to the service or not. Make sure Azure Monitor is configured to monitor the service before starting the migration (for more info on how to configure Azure monitor, please refer to [the Azure Monitor guide](../how-to-guides/configure-dicom-server-settings.md#azure-monitor)).
```

    3.1. If you are ok with interruption to the service, you can follow the steps below:

>         3.1.1. Change the BlobMigration.StartCopy to "True" and restart the service. This will start the copy background service which will trigger the CopyFiles Azure Durable Function which will copy the old format DICOM files to new format.

Reviewer: We should add something in section 3 about Durable Functions setup or link to existing documentation.

```suggestion
        3.1.1. Set `BlobMigration.StartCopy` to `true` and restart the service. This will trigger the `CopyFiles` Durable Function which will copy the old format DICOM files to the new format.
```

Author: I was planning but there is no setup. If they deploy through our ARM template, it should be deployed already. Only thing they need to know is about Azure Monitor to look out for logs.

        3.1.2. To ensure Copy has been completed, you can check Azure Monitor logs for `"Completed copying files."` message. This will indicate that the copy has been completed.

>         3.1.3. Once the copy is completed, you can change the BlobMigration.FormatType to "New" and BlobMigration.StartDelete to "True" and restart the service. This will trigger Delete background service which will delete all the old format blobs only if the corresponding new format blobs exist and set the format to New. This is a safe operation and doesn't delete any blobs without checking for the existence of new format blobs.

Reviewer: ```suggestion
        3.1.3. Once the copy is completed, you can change `BlobMigration.FormatType` to `"New"` and `BlobMigration.StartDelete` to `true` and restart the service. This will trigger a Durable Function which will delete all the old format blobs only if corresponding new format blobs exist. This is a safe operation and doesn't delete any blobs without checking for the existence of new format blobs.
```

        3.1.4. To ensure Delete has been completed, you can check Azure Monitor logs for `"Completed deleting files."` message. This will indicate that the delete has been completed.

    3.2. If you are not ok with interruption to the service, you can follow the steps below:

>         3.2.1. Change the BlobMigration.FormatType to "Dual" and restart the service. This will duplicate any new DICOM files uploaded to both old and new format. 

Reviewer: ```suggestion
        3.2.1. Change `BlobMigration.FormatType` to `"Dual"` and restart the service. This will duplicate any new DICOM files uploaded to both old and new format.
```

>         3.2.2 Follow steps 3.1.1 to 3.1.4 to complete the copy and delete operation.

Author: We need the dualwrite step to run the service without interruption. Once they upgrade, it will all be dualwrite. It is a stopgap solution

>         3.2.2 Follow steps 3.1.1 to 3.1.4 to complete the copy and delete operation.

Reviewer: Step 3.2.2 says to perform step 3.1.4 which will set the format to `New`

#### With Service Interruption
If you are ok with interruption to the service, you can follow the steps below. The interruption here is a self-managed one and using the service while the copying is occuring and before switching to the new format can corrupt the data path or retrieve wrong data.

1. Set `BlobMigration.StartCopy` to `true` and restart the service.
   1. When restarting the service, ensure it is in such a way that it picks up the new application settings and this will vary by how your service is deployed.
   2. This will trigger the `CopyFiles` Durable Function which will copy the old format DICOM files to the new format.
   3**Do not use the service at this time**. We want to make sure all files are copied over.

> 2. To ensure Copy has been completed, you can check Azure Monitor logs for `"Completed copying files."` message. This will indicate that the copy has been completed.

Reviewer: ```suggestion
2. To ensure the Copy operation has been completed, you can check Azure Monitor logs for a `"Completed copying files."` message. This will indicate that the operation has been completed:
```

   ![dicomwebcontainer-bluecircle-copy-logs](../images/dicomwebcontainer-bluecircle-copy-logs.png)

>     At this time, you'll have both the new and old files

Reviewer: ```suggestion
    At this time, you'll have both the new and old files:
```

    ![dicomwebcontainer-bluecircle-old-blob-format-dual](../images/dicomwebcontainer-bluecircle-old-blob-format-dual.png)

3. Once the copy is completed, you can change `BlobMigration.FormatType` to `"New"` and `BlobMigration.StartDelete` to `true` and restart the service.
   1. This will trigger a Durable Function which will delete all the old format blobs only if corresponding new format blobs exist. This is a safe operation and doesn't delete any blobs without checking for the existence of new format blobs.
   2. **Start using your service** to store and retrieve data, which will work with the `New` format.
4. To ensure Delete has been completed, you can check Azure Monitor logs for `"Completed deleting files."` message. This will indicate that the delete has been completed.

#### Without Service Interruption
If you are not ok with interruption to the service, you can follow the steps below.

> 1. Change `BlobMigration.FormatType` to `"Dual"`

Reviewer: There only need to be two steps here - let's combine 1. and sub-1.

   1. This will duplicate any new DICOM files uploaded to both old and new format as you continue to use your service during the copying operation.
2. Follow steps in [With Service Interruption](#with-service-interruption), but feel free to continue using the service.

> **Note: If in case there are any issues or questions during migration, you can post your questions or issues in the [DICOM Server GitHub Discussions](https://github.com/microsoft/dicom-server/discussions/1561) page.**`

Reviewer: ```suggestion
> **Please post any questions or issues encountered during migration in the related [GitHub Discussion](https://github.com/microsoft/dicom-server/discussions/1561).**
```
